File: isPalindrome/isPalindrome.cpp

#include "../include/include.cpp"

class Solution {
public:
    bool isPalindrome(string s) {

        string str = characterizeString(s);
        int len = str.size();
        for (int i = 0; i < len / 2; i++)
        {
            if (str[i] != str[len - i - 1])
                return false;
        }

        return true;
    }


    string characterizeString(string s)
    {
        string n;
        for (int i = 0; s[i]; i++)
        {
            if (s[i] >= 97 && s[i] <= 122)
                n.push_back(s[i]);
            else if (s[i] >= 65 && s[i] <= 90)
                n.push_back(s[i] + 32);
            else if (s[i] >= '0' && s[i] <= '9')
                n.push_back(s[i]);
        }
        return n;
    }
};


int main() {


    Solution ss = Solution();

    vector<string> in = {
        "A man, a plan, a canal: Panama",
        "race a car",
        " "
    };

    for (auto i : in)
    {
        cout << ss.isPalindrome(i) << '\n';
    }


    return 0;
}

File: isPalindrome/isPalindrome.py


class Solution:
    def isPalindrome(self, s):

        start = 0
        end = len(s) - 1

        while start < end:

            while start < end and not self.isalphanum(s[start]):
                start += 1
            while start < end and not self.isalphanum(s[end]):
                end -= 1
            if s[start].lower() == s[end].lower():
                start += 1
                end -= 1
            else:
                return False

        return True

    def isalphanum(self, c):
        return (ord('A') <= ord(c) <= ord('Z') or
                ord('a') <= ord(c) <= ord('z') or
                ord('0') <= ord(c) <= ord('9'))

        






ss = Solution()


print(ss.isPalindrome("A man, a plan, a canal: Panama"))

